From 6c875c59d5375294afb6dcc719f93d679102ebbc Mon Sep 17 00:00:00 2001
From: Surge Raval <Surge1223@gmail.com>
Date: Wed, 4 Jan 2017 10:29:34 -0800
Subject: [PATCH 1/2] Introduce sepolicy exceptions for theme assets

Assets such as composed icons and ringtones need to be accessed
by apps. This patch adds the policy needed to facilitate this.

commit eb25c1e0b69aab8d0c055d51de15501827f5dffd
Author: George G <kreach3r@users.noreply.github.com>
Date:   Wed Feb 8 17:22:44 2017 +0200

    sepolicy: fix themed sounds

    02-08 17:26:48.011 18259-18259/? W/SoundPoolThread: type=1400 audit(0.0:31): avc: denied { read } for path="/data/system/theme/audio/ui/Lock.ogg" dev="dm-0" ino=1006317 scontext=u:r:drmserver:s0 tcontext=u:object_r:theme_data_file:s0 tclass=file permissive=0

commit 4b1a5419b93858200e9cc143c900fc02ad2781d4
Author: Miccia <bono.michele94@gmail.com>
Date:   Mon Feb 27 12:36:21 2017 +0100

    sepolicy: Fix application of bootanimation

commit 2af86e0a49b496da3106885e5c6aa1c23e99db9d
Author: Randall Rushing <randall.rushing@gmail.com>
Date:   Wed Jan 4 10:31:29 2017 -0800

    sepolicy: fix themed boot animation

    W BootAnimation: type=1400 audit(0.0:42): avc: denied { open } for uid=1003 path="/data/system/theme/bootanimation.zip" dev="mmcblk0p42" ino=1657697 scontext=u:r:bootanim:s0 tcontext=u:object_r:system_data_file:s0 tclass=file permissive=0

    W         : Unable to open '/data/system/theme/bootanimation.zip': Permission denied

    W zipro   : Error opening archive /data/system/theme/bootanimation.zip: I/O Error

Change-Id: I0420de579aed0cff5add181cd0a8bf0f2b05d723
Signed-off-by: Harsh Shandilya <msfjarvis@gmail.com>
---
 private/app.te           | 4 ++++
 private/file_contexts    | 3 +++
 private/system_app.te    | 3 +++
 private/system_server.te | 3 +++
 private/zygote.te        | 4 ++++
 public/bootanim.te       | 5 +++++
 public/drmserver.te      | 4 ++++
 public/file.te           | 3 +++
 public/mediaserver.te    | 4 ++++
 9 files changed, 33 insertions(+)

diff --git a/private/app.te b/private/app.te
index 5b077ac..6fdef15 100644
--- a/private/app.te
+++ b/private/app.te
@@ -523,3 +523,7 @@ neverallow {
   -bluetooth
   -system_app
 } bluetooth_prop:file create_file_perms;
+
+# Themed resources (i.e. composed icons)
+allow appdomain theme_data_file:dir r_dir_perms;
+allow appdomain theme_data_file:file r_file_perms;
diff --git a/private/file_contexts b/private/file_contexts
index 4485b95..901e9ed 100644
--- a/private/file_contexts
+++ b/private/file_contexts
@@ -394,6 +394,9 @@
 # Bootchart data
 /data/bootchart(/.*)?		u:object_r:bootchart_data_file:s0
 
+# Themes data
+/data/system/theme(/.*)?    u:object_r:theme_data_file:s0
+
 #############################
 # Expanded data files
 #
diff --git a/private/system_app.te b/private/system_app.te
index 7950044..e8b801a 100644
--- a/private/system_app.te
+++ b/private/system_app.te
@@ -81,6 +81,9 @@ allow system_app keystore:keystore_key {
 # /sys access
 r_dir_file(system_app, sysfs_type)
 
+# /data/system access
+allow system_app system_data_file:file r_file_perms;
+
 control_logd(system_app)
 read_runtime_log_tags(system_app)
 
diff --git a/private/system_server.te b/private/system_server.te
index 90bc57e..be9ef4f 100644
--- a/private/system_server.te
+++ b/private/system_server.te
@@ -669,6 +669,9 @@ with_asan(`
   allow system_server shell_exec:file rx_file_perms;
 ')
 
+# allow system_server to look into theme resources
+allow system_server theme_data_file:dir search;
+
 ###
 ### Neverallow rules
 ###
diff --git a/private/zygote.te b/private/zygote.te
index daabbc0..4cd84da 100644
--- a/private/zygote.te
+++ b/private/zygote.te
@@ -112,6 +112,10 @@ allow zygote tmpfs:dir r_dir_perms;
 # Let the zygote access overlays so it can initialize the AssetManager.
 get_prop(zygote, overlay_prop)
 
+# Themes
+allow zygote theme_data_file:file r_file_perms;
+allow zygote theme_data_file:dir r_dir_perms;
+
 ###
 ### neverallow rules
 ###
diff --git a/public/bootanim.te b/public/bootanim.te
index e2584c3..7b911f4 100644
--- a/public/bootanim.te
+++ b/public/bootanim.te
@@ -38,3 +38,8 @@ r_dir_file(bootanim, cgroup)
 
 # System file accesses.
 allow bootanim system_file:dir r_dir_perms;
+
+# Themed resources (bootanimation)
+allow bootanim theme_data_file:dir search;
+allow bootanim theme_data_file:file r_file_perms;
+allow bootanim system_data_file:file open;
diff --git a/public/drmserver.te b/public/drmserver.te
index f752c13..80d6403 100644
--- a/public/drmserver.te
+++ b/public/drmserver.te
@@ -56,3 +56,7 @@ selinux_check_access(drmserver)
 
 r_dir_file(drmserver, cgroup)
 r_dir_file(drmserver, system_file)
+
+# Themed resources (i.e. composed icons)
+allow drmserver theme_data_file:dir r_dir_perms;
+allow drmserver theme_data_file:file r_file_perms;
diff --git a/public/file.te b/public/file.te
index 057af41..87b882f 100644
--- a/public/file.te
+++ b/public/file.te
@@ -335,3 +335,6 @@ allow postinstall_file self:filesystem associate;
 # Should be:
 #   type apk_data_file, file_type, data_file_type;
 neverallow fs_type file_type:filesystem associate;
+
+# Themes
+type theme_data_file, file_type, data_file_type;
diff --git a/public/mediaserver.te b/public/mediaserver.te
index 6efaf0f..9f621bd 100644
--- a/public/mediaserver.te
+++ b/public/mediaserver.te
@@ -148,3 +148,7 @@ neverallow mediaserver { file_type fs_type }:file execute_no_trans;
 
 # do not allow privileged socket ioctl commands
 neverallowxperm mediaserver domain:{ rawip_socket tcp_socket udp_socket } ioctl priv_sock_ioctls;
+
+# Themed resources (i.e. composed icons)
+allow mediaserver theme_data_file:dir r_dir_perms;
+allow mediaserver theme_data_file:file r_file_perms;
-- 
2.1.2

